### ISSUE TITLE
engine truncating method calls in interpolation in attributes

### ISSUE THREAD

**Issue by michaelglass** [k0kubun/hamlit#118]
[Here's a minimal test case](https://github.com/NoRedInk/hamlit/commit/254c1e04a30cd7de9e4b2fb084e56dcdc309dc32).  I don't have Temple or C chops to fix this totally on my own. Happy to dive in if anyone wants to pair!

Does _not_ happen, (e.g. everything is good) if...
- method calls have parens
    ```haml
    %p{class: "#{"1".concat("2", "3")}"
    ```
- if it's just string interpolation in text nodes
    ```haml
    %p "#{"1".concat "2", "3"}"
    ```

fails when
- string interpolation
- with a method / function call
- with multiple params
- without parens around params
    ```haml
    %p{class: "#{"1".concat "2", "3"}"
    ```

```ruby
def test_interpolation_should_work_inside_attributes
  assert_equal("123", "#{"1".concat "2", "3"}")
  assert_equal("<p class='a 13'>foo</p>", render('%p{class: "a #{"1".concat "2", "3"}"} foo').chomp)
end
```

produces

```shell
Error:
EngineTest#test_interpolation_should_work_inside_attributes:
SyntaxError: (eval):1: syntax error, unexpected ')', expecting tSTRING_DEND
..._class(true, "#{"1".concat "2")); _buf << ("'></p>\n".freeze...
...                              ^
(eval):1: syntax error, unexpected end-of-input, expecting tSTRING_DEND
...e); _buf = _buf.join("".freeze)
...                               ^
    test/test_helper.rb:62:in `eval'
    test/test_helper.rb:62:in `render'
    test/haml/engine_test.rb:108:in `render'
    test/haml/engine_test.rb:154:in `test_interpolation_should_work_inside_attributes'```

the string that's produced is

```ruby
_buf = []; _buf << ("<p 3='".freeze); _buf << (::Hamlit::Utils.escape_html(())); _buf << ("' class='".freeze); _buf << (::Hamlit::AttributeBuilder.build_class(true, "#{"1".concat "2")); _buf << ("'></p>\n".freeze); _buf = _buf.join("".freeze)
```

hand-prettified, that's
```ruby
_buf = []
_buf << "<p 3='".freeze
_buf << ::Hamlit::Utils.escape_html(())
_buf << "' class='".freeze
#                                            missing stuff here     vvvv
_buf << ::Hamlit::AttributeBuilder.build_class(true, "#{"1".concat "2")
#                                            missing stuff here     ^^^^
_buf << "'></p>\n".freeze
_buf = _buf.join("".freeze)
```

**Comment by WaKeMaTTa** [comment]
```diff
- assert_equal("<p class='a 13'>foo</p>", render('%p{class: "a #{"1".concat "2", "3"}"} foo').chomp)
+ assert_equal("<p class='a 123'>foo</p>", render('%p{class: "a #{"1".concat "2", "3"}"} foo').chomp)
```

**Comment by k0kubun** [comment]
Fixed in Hamlit 2.8.7. Please try it.

**Comment by michaelglass** [comment]
I have never had someone address a bug report so quickly. I have been telling everyone I know about this experience (including people that have no idea what I'm talking about, e.g. my friends' children).

Thank you so much!

![](https://media.giphy.com/media/xTiTnEHBh7qapyuvwQ/giphy.gif)
